### DISCUSSION TITLE
Setting cookies?

### DISCUSSION THREAD

**Discussion by peace2000** [uBlockO/uBO-Scriptlets#21] [Q&A] [answered]
Hi @uBlock-user, as you have scriplets to set local items and session items, could you also consider adding a scriplet that allows to set cookies on websites? This could be useful as it'd allow automatically to force specific settings on websites or restore them if the browser cache has been emptied.

One use case: if you set a cookie on Youtube `https://www.youtube.com/` which Name is `SOCS` and Value is `CAESEwgDEgk1MjE4NjcxMTIaAmVuIAEaBgiAl7ihBg`, that would allow to skip the GDPR dialog automatically with the setting "Reject all".

<details>

![kuva](https://user-images.githubusercontent.com/17256841/230494050-d9ed7cb0-8996-4ed5-be03-a0e2db001e57.png)

</details>

**Comment by uBlock-user** [comment, marked as answer]
```
/// setCookie.js
// example.com##+js(setCookie, SOCS, CAESEwgDEgk1MjE4NjcxMTIaAmVuIAEaBgiAl7ihBg)
function( 
	cName = '', 
	cValue = '',
	expTime = '',
	behavior = '' 
) {
	let timer = undefined;
	if ( cName === '' ) { return; }
	if ( expTime === '' ) { expTime = '1'; }
	if ( cValue === '' ) { cValue = '""'; }	
	const setCookie = () => {
		try {
			const cookies = document.cookie;
			if ( !cookies.includes(cName) || !cookies.includes(cValue) ) {
		      		const date = new Date();
				date.setTime(date.getTime() + (expTime * 24 * 60 * 60 * 1000));
				const expires = "expires=" + date.toUTCString();
				document.cookie = cName + "=" + cValue + "; " + expires + "; path=/";
			} else { 
				return; 
			}	
		} catch { }
	};
	const mutationHandler = mutations => {
	if ( timer !== undefined ) { return; }
	let skip = true;
	for ( let i = 0; i < mutations.length && skip; i++ ) {
	    const { type, addedNodes, removedNodes } = mutations[i];
	    if ( type === 'attributes' ) { skip = false; }
	    for ( let j = 0; j < addedNodes.length && skip; j++ ) {
		if ( addedNodes[j].nodeType === 1 ) { skip = false; break; }
	    }
	    for ( let j = 0; j < removedNodes.length && skip; j++ ) {
		if ( removedNodes[j].nodeType === 1 ) { skip = false; break; }
	    }
	}
	if ( skip ) { return; }
	timer = self.requestIdleCallback(setCookie, { timeout: 10 });
	};
	const start = ( ) => {
	setCookie();
	if ( /\bloop\b/.test(behavior) === false ) { return; }
	const observer = new MutationObserver(mutationHandler);
	observer.observe(document.documentElement, {
	    attributes: true,
	    childList: true,
	    subtree: true,
	});
	};
	if ( document.readyState !== 'complete' && /\bcomplete\b/.test(behavior) ) {
	self.addEventListener('load', start, { once: true });
	} else if ( document.readyState !== 'loading' || /\basap\b/.test(behavior) ) {
	start(); 
	} else {
	self.addEventListener('DOMContentLoaded', start, { once: true });
	}	
}
   
```

@peace2000

**Reply by peace2000** [reply]
Hi, thanks! This script is definetely useful on many websites but it seems that Youtube checks whether that GDPR cookie is there or not, faster than the cookie is injected by using this scriplet, which leads the GDPR dialog to be shown unfortunately.

Could you add `asap` option to this?

**Reply by uBlock-user** [reply]
Did you not use `loop` first ?

**Reply by peace2000** [reply]
Hi, tested `loop` but it doesn't change anything.

Your script does inject the cookie, but the site is quicker to check the cookie's existance than the script injects it.

When I refresh the page, the GDPR dialog disappears because of the injected cookie, so the cookie itself works but it's injected too slowly.

**Reply by uBlock-user** [reply]
Added the `asap` in the code, import the above scriptlet and try again.

**Reply by peace2000** [reply]
Hmm I tried to lower `{ timeout: 10 });` to `{ timeout: 9 });` but now creating that cookie stopped working completely. Probably 9 is too low value.

**Reply by uBlock-user** [reply]
So `asap` doesn't work either ?

**Reply by peace2000** [reply]
Is this the correct way? I tried to add `asap` similarly as it's in some uBO's built-in scriplets.

<details>

```
'use strict';

/// setCookie.js
/// alias sck.js
// example.com##+js(sck, cookie, value)
function( 
	cName = '', 
	cValue = '',
	expTime = '',
	behavior = '' 
) {
	let timer = undefined;
	if ( cName === '' ) { return; }
	if ( expTime === '' ) { expTime = '1'; }
	if ( cValue === '' ) { cValue = '""'; }	
	const setCookie = () => {
		try {
			const cookies = document.cookie;
			if ( !cookies.includes(cName) || !cookies.includes(cValue) ) {
		      		const date = new Date();
				date.setTime(date.getTime() + (expTime * 24 * 60 * 60 * 1000));
				const expires = "expires=" + date.toUTCString();
				document.cookie = cName + "=" + cValue + "; " + expires + "; path=/";
			} else { 
				return; 
			}	
		} catch { }
	};
	const mutationHandler = mutations => {
	if ( timer !== undefined ) { return; }
	let skip = true;
	for ( let i = 0; i < mutations.length && skip; i++ ) {
	    const { type, addedNodes, removedNodes } = mutations[i];
	    if ( type === 'attributes' ) { skip = false; }
	    for ( let j = 0; j < addedNodes.length && skip; j++ ) {
		if ( addedNodes[j].nodeType === 1 ) { skip = false; break; }
	    }
	    for ( let j = 0; j < removedNodes.length && skip; j++ ) {
		if ( removedNodes[j].nodeType === 1 ) { skip = false; break; }
	    }
	}
	if ( skip ) { return; }
	timer = self.requestIdleCallback(setCookie, { timeout: 9 });
	};
	const start = ( ) => {
	setCookie();
	if ( /\bloop\b/.test(behavior) === false ) { return; }
	const observer = new MutationObserver(mutationHandler);
	observer.observe(document.documentElement, {
	    attributes: true,
	    childList: true,
	    subtree: true,
	});
	};
	if ( document.readyState !== 'complete' && /\bcomplete\b/.test(behavior) ) {
	self.addEventListener('load', start, { once: true });
	} else if ( document.readyState !== 'loading' || /\basap\b/.test(behavior) ) {
	start();
	} else {
		self.addEventListener('DOMContentLoaded', start, { once: true });
    }	
}
```

</details>

I'll have  to test this more with various different timeout values.

**Reply by uBlock-user** [reply]
The filter would be `example.com##+js(sck, cookie, value, time, asap)`

**Reply by uBlock-user** [reply]
Feel free to test this version too --

```
/// test.js
// example.com##+js(test, cookie, value)
function( 
	cName = '', 
	cValue = '',
	expTime = '' 
) {
	if ( cName === '' ) { return; }
	if ( expTime === '' ) { expTime = '1'; }
	if ( cValue === '' ) { cValue = '""'; }	
	const setCookie = () => {
		try {
			const cookies = document.cookie;
			if ( !cookies.includes(cName) || !cookies.includes(cValue) ) {
		      		const date = new Date();
				date.setTime(date.getTime() + (expTime * 24 * 60 * 60 * 1000));
				const expires = "expires=" + date.toUTCString();
				document.cookie = cName + "=" + cValue + "; " + expires + "; path=/";
			} else { 
				return; 
			}	
		} catch { }
	};
	const observer = new MutationObserver(setCookie);
	observer.observe( document.documentElement, { childList: true, subtree: true } );
	if ( document.readyState === "complete" ) { observer.disconnect(); }
}
```

**Reply by peace2000** [reply]
Thanks.

In order to skip that GDPR dialog on Youtube, `asap` needs to be in value `8` (`9` didn't work fast enough, didn't test values between those two):

`youtube.com##+js(sck, SOCS, CAESEwgDEgk1MjE4NjcxMTIaAmZpIAEaBgiA3cKhBg, 8, asap)`

That latter scriptlet worked without `asap`:

`youtube.com##+js(test, SOCS, CAESEwgDEgk1MjE4NjcxMTIaAmZpIAEaBgiA3cKhBg)`

With Youtube, a long-term solution doesn't seem to be possible as I noticed that cookie value changes... So this is useful for sites that don't change their cookie values.

**Reply by uBlock-user** [reply]
@peace2000 I made a mistake with `asap` implementation please try the previous scriptlet I modified again. If it still doesn't work, then I will change timeout value to `5`.

**Reply by uBlock-user** [reply]
> 8

Third parameter is cookie expiry time, so it has nothing to do with race condition.

**Reply by peace2000** [reply]
Oh ok.

Tested now, both work now just fine, no race-condition.

Btw I noticed a strange thing: I cannot set any cookies on https://www.google.fi/ if I'm using Google Chrome... Do you have any idea why?

**Reply by uBlock-user** [reply]
`asap` version has now been added to uBO-Scriptlets repository. 

> Btw I noticed a strange thing: I cannot set any cookies on https://www.google.fi/ if I'm using Google Chrome... Do you have any idea why?

I don't have Chrome installed so cannot tell.

**Reply by peace2000** [reply]
Ok scrap my previous comment concerning google.fi, it was just a random hiccup probably, now it works.

Thanks for this scriplet!
